blockchain/chain_test.go

```diff
@@ -1311,3 +1313,381 @@ func TestIsAncestor(t *testing.T) {
 			branch2Nodes[0].hash.String())
 	}
 }
+
+func TestInvalidateBlock(t *testing.T) {
+	tests := []struct {
+		name     string
+		chainGen func() (*BlockChain, []*chainhash.Hash, func())
+	}{
+		{
+			name: "One branch, invalidate once",
+			chainGen: func() (*BlockChain, []*chainhash.Hash, func()) {
+				source := rand.NewSource(time.Now().UnixNano())
+				rand := rand.New(source)
+
+				chain, params, tearDown := utxoCacheTestChain("TestInvalidateBlock-one-branch-invalidate-once")
+				tip := btcutil.NewBlock(params.GenesisBlock)
+
+				// Create block at height 1.
+				var emptySpendableOuts []*testhelper.SpendableOut
+				b1, spendableOuts1, err := addBlock(chain, tip, emptySpendableOuts)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				var allSpends []*testhelper.SpendableOut
+				nextBlock := b1
+				nextSpends := spendableOuts1
+
+				var invalidateHash *chainhash.Hash
+
+				// Create a chain with 11 blocks.
+				for b := 0; b < 10; b++ {
+					newBlock, newSpendableOuts, err := addBlock(chain, nextBlock, nextSpends)
+					if err != nil {
+						t.Fatal(err)
+					}
+
+					nextBlock = newBlock
+
+					if newBlock.Height() == 5 {
+						invalidateHash = newBlock.Hash()
+					}
+
+					allSpends = append(allSpends, newSpendableOuts...)
+
+					var nextSpendsTmp []*testhelper.SpendableOut
+					for i := 0; i < len(allSpends); i++ {
```

Comment:
A comment explaining what this is doing would be very helpful

Reply:
Addressed in the latest pushes.